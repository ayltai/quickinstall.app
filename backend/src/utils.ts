import { ErrorReporting, } from '@google-cloud/error-reporting';

const errors = new ErrorReporting();

export const handleError = (error : any) => {
    if (process.env.NODE_ENV === 'production') {
        errors.report(error);
    } else {
        console.error(error);
    }
};
