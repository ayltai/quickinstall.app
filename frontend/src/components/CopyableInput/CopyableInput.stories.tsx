import type { Meta, StoryObj, } from '@storybook/react';

import { CopyableInput, } from './CopyableInput';

const meta : Meta<typeof CopyableInput> = {
    component : CopyableInput,
    title     : 'Components/Inputs/Copyable Input',
    argTypes  : {
        value  : {
            description : 'The value to copy',
        },
        onCopy : {
            description : 'Function to call when the copy button is clicked',
        },
    },
};

export default meta;

type Story = StoryObj<typeof CopyableInput>;

export const Default : Story = {
    args : {
        value : 'This is a test',
    },
};
