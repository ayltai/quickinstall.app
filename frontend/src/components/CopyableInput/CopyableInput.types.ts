import type { InputBaseProps, } from '@mui/material/InputBase';

export interface CopyableInputProps extends InputBaseProps {
    value   : string,
    onCopy? : () => void,
}
