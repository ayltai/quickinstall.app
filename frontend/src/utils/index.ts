import * as Sentry from '@sentry/react';
import { FirebaseApp, } from 'firebase/app';
import { getAnalytics, logEvent, } from 'firebase/analytics';

export const handleError = (error : any) => {
    if (import.meta.env.PROD) {
        Sentry.captureException(error);
    } else {
        console.error(error);
    }
};

export const logPageView = (app : FirebaseApp, {
    pageName,
    route,
} : {
    pageName : string,
    route    : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'page_view', {
            page_title    : pageName,
            page_location : route,
        });
    } catch (error) {
        handleError(error);
    }
};

export const logAddToCart = (app : FirebaseApp, {
    platform,
    itemId,
} : {
    platform : 'win' | 'mac',
    itemId   : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'add_to_cart', {
            items : [
                {
                    item_id : `${platform}/${itemId}`,
                },
            ],
        });
    } catch (error) {
        handleError(error);
    }
};

export const logRemoveFromCart = (app : FirebaseApp, {
    platform,
    itemId,
} : {
    platform : 'win' | 'mac',
    itemId   : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'remove_from_cart', {
            items : [
                {
                    item_id : `${platform}/${itemId}`,
                },
            ],
        });
    } catch (error) {
        handleError(error);
    }
};

export const logCheckout = (app : FirebaseApp, {
    platform,
    itemIds,
} : {
    platform : 'win' | 'mac',
    itemIds  : string[],
}) => {
    try {
        logEvent(getAnalytics(app), 'checkout', {
            items : itemIds.map((itemId : string) => ({
                item_id : `${platform}/${itemId}`,
            })),
        });
    } catch (error) {
        handleError(error);
    }
};

export const logSearch = (app : FirebaseApp, {
    platform,
    query,
} : {
    platform : 'win' | 'mac',
    query    : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'search', {
            search_term : `${platform}/${query}`,
        });
    } catch (error) {
        handleError(error);
    }
};

export const logShare = (app : FirebaseApp, {
    product,
    itemId,
} : {
    product : 'win/packages' | 'mac/packages' | 'tweaks',
    itemId  : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'share', {
            content_type : product,
            item_id      : itemId,
        });
    } catch (error) {
        handleError(error);
    }
};
