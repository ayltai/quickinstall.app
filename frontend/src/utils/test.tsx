import { createTheme, ThemeProvider, } from '@mui/material';
import { configureStore, type Store, } from '@reduxjs/toolkit';
import { render, } from '@testing-library/react';
import React, { type ReactNode, } from 'react';
import { Provider, } from 'react-redux';
import { MemoryRouter, } from 'react-router-dom';

import { appStoreApi, chocolateyApi, contentfulApi, homebrewApi, quickInstallApi, } from '../apis';
import { recipeReducer, preferenceReducer, spellReducer, } from '../states';
import { appTheme, } from '../styles';

export const createStore = (preloadedState? : unknown) => configureStore({
    preloadedState,
    reducer    : {
        recipe                          : recipeReducer,
        preference                      : preferenceReducer,
        spell                           : spellReducer,
        [ appStoreApi.reducerPath  ]    : appStoreApi.reducer,
        [ chocolateyApi.reducerPath ]   : chocolateyApi.reducer,
        [ contentfulApi.reducerPath ]   : contentfulApi.reducer,
        [ homebrewApi.reducerPath  ]    : homebrewApi.reducer,
        [ quickInstallApi.reducerPath ] : quickInstallApi.reducer,
    },
    middleware : getDefaultMiddleware => getDefaultMiddleware().concat(appStoreApi.middleware, chocolateyApi.middleware, contentfulApi.middleware, homebrewApi.middleware, quickInstallApi.middleware),
});

export const defaultStore = createStore();

export type RootState = ReturnType<typeof defaultStore.getState>;

// @ts-ignore
const theme = createTheme(appTheme);

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : Record<string, any>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => (
        <Provider store={store}>
            <MemoryRouter>
                <ThemeProvider theme={theme}>
                    {children}
                </ThemeProvider>
            </MemoryRouter>
        </Provider>
    ),
    ...rest,
});

export * from '@testing-library/react';

export { customRender as render, };
