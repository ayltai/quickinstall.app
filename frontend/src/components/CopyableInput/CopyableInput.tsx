import ContentCopy from '@mui/icons-material/ContentCopy';
import IconButton from '@mui/material/IconButton';
import InputAdornment from '@mui/material/InputAdornment';
import OutlinedInput from '@mui/material/OutlinedInput';
import Snackbar from '@mui/material/Snackbar';
import Tooltip from '@mui/material/Tooltip';
import { styled, } from '@mui/material/styles';
import React, { type FC, type FocusEvent, useState, } from 'react';
import { useTranslation, } from 'react-i18next';

import { handleError, } from '../../utils';
import type { CopyableInputProps, } from './CopyableInput.types';

const MonospaceInput = styled(OutlinedInput)({
    paddingLeft  : '0.5rem',
    paddingRight : '1.5rem',
    borderRadius : '2rem',
    fontFamily   : 'monospace',
});

/**
 * An input field that allows a user to copy information to the clipboard.
 * @param value The value to copy
 * @param onCopy Function to call when the copy button is clicked
 * @param rest Other props
 */
export const CopyableInput : FC<CopyableInputProps> = ({
    value,
    onCopy,
    ...rest
}) => {
    const [ open, setOpen, ] = useState(false);

    const { t, } = useTranslation();

    const handleCopy = () => {
        if (value) navigator.clipboard.writeText(value).catch(handleError);

        setOpen(true);

        if (onCopy) onCopy();
    };

    const handleFocus = (event : FocusEvent<HTMLInputElement>) => event.target.select();

    const handleClose = () => setOpen(false);

    return (
        <>
            <MonospaceInput
                {...rest}
                value={value}
                endAdornment={navigator.clipboard && (
                    <InputAdornment position='end'>
                        <Tooltip title={t('action.copy')}>
                            <IconButton
                                edge='end'
                                onClick={handleCopy}>
                                <ContentCopy />
                            </IconButton>
                        </Tooltip>
                    </InputAdornment>
                )}
                onFocus={handleFocus} />
            <Snackbar
                open={open}
                anchorOrigin={{
                    horizontal : 'center',
                    vertical   : 'bottom',
                }}
                autoHideDuration={2500}
                message={t('common.copied')}
                onClose={handleClose} />
        </>
    );
};
