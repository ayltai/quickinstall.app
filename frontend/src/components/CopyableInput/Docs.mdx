import { Controls, Description, Meta, Title, Primary, Stories, } from '@storybook/blocks';

import * as ComponentStories from './CopyableInput.stories';

<Meta of={ComponentStories} />

<Title />

<Description />

# How to use copyable inputs

Copyable inputs are used to allow users to copy data.

When using a copyable input,
* Use a label to explain exactly what can be done with it

# Default usage

<Primary />

# Parameters

<Controls sort='requiredFirst' />

<Stories
    title='Variants of buttons'
    includePrimary={false} />
