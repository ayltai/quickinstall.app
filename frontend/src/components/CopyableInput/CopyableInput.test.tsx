import React from 'react';

import { fireEvent, render, waitFor, } from '../../utils/test';
import { CopyableInput, } from './CopyableInput';

const DUMMY_VALUE = 'This is a test';

const component = (
    <CopyableInput value={DUMMY_VALUE} />
);

const writeText = vi.fn(() => Promise.resolve());

Object.assign(navigator, {
    clipboard : {
        writeText,
    },
});

describe('<CopyableInput />', () => {
    it('renders correctly', () => expect(render(component)).toMatchSnapshot());

    it('copies input value to clipboard', async () => {
        const { getByRole, } = render(component);

        fireEvent.click(getByRole('button'));

        await waitFor(() => {
            expect(writeText).toHaveBeenCalledWith(DUMMY_VALUE);
        });
    });
});
