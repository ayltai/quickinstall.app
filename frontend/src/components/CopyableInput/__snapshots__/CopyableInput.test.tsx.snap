// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<CopyableInput /> > renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-adornedEnd css-16tmvdj-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
          type="text"
          value="This is a test"
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd css-1laqsz7-MuiInputAdornment-root"
        >
          <button
            aria-label="action.copy"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
            data-mui-internal-clone-element="true"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ContentCopyIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2m0 16H8V7h11z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-ihdtdm"
          >
            <span
              class="notranslate"
            >
              ​
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-adornedEnd css-16tmvdj-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
        type="text"
        value="This is a test"
      />
      <div
        class="MuiInputAdornment-root MuiInputAdornment-positionEnd css-1laqsz7-MuiInputAdornment-root"
      >
        <button
          aria-label="action.copy"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element="true"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ContentCopyIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2m0 16H8V7h11z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-ihdtdm"
        >
          <span
            class="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
